### ISSUE TITLE
Validate DOM nesting when possible

### ISSUE THREAD

**Issue by karlhorky** [jsx-eslint/eslint-plugin-react#3310]
Hey @ljharb, long time no see! 👋 Have been pretty inactive around here, sorry about that! Lots of new students.

Wanted to see what you think about a new feature to validate DOM nesting when possible, to avoid such errors as seen below - which now, as of React 18, lead to hydration errors.

**Problem:** `th` is not a valid child of `tfoot`

```jsx
export default function IndexPage() {
  return (
    <div>
      <table>
        <tfoot>
          <th>abc</th> {/* 💥 Error: Hydration failed because the initial UI does not match what was rendered on the server. */}
        </tfoot>
      </table>
    </div>
  );
}
```

I remember React did some kind of simple validation of DOM nesting, although I haven't seen this error for a while now (maybe just haven't run into it any more).

Maybe `eslint-plugin-react` could have some simple JSX validation (maybe based on a well-maintained library). Not across component boundaries or anything, only static React elements which are HTML tags.

It leads to this (pretty confusing) error with latest Next.js (see [this StackBlitz](https://stackblitz.com/edit/nextjs-mcaiql?file=package.json,pages%2Findex.js)):

![Screen Shot 2022-06-15 at 16 44 45](https://user-images.githubusercontent.com/1935696/173857859-f419e8e9-be49-4a60-b6bd-24dcfb486979.png)

If you were to open the DevTools and scroll up all the way to the top, you would see a more helpful error message (which, arguably, React should also display), but it's pretty hidden:

![Screen Shot 2022-06-15 at 16 45 15](https://user-images.githubusercontent.com/1935696/173858096-e52b53f5-c9f2-443b-91d7-5a281ac0d571.png)

**Comment by ljharb** [comment]
I think that this would be very useful IFF we had a shared data source between React and this plugin, that matched what browsers accepted. Do you know of one?

**Comment by karlhorky** [comment]
There's this file from the React source here: https://github.com/facebook/react/blob/main/packages/react-dom/src/client/validateDOMNesting.js (apparently Sophie extracted a subset of the spec, Dan mentioned)

Maybe that plus some additional rules for common things?

**Comment by ljharb** [comment]
Is that reachable from the actual npm package?

**Comment by karlhorky** [comment]
Seems like `validateDOMNesting` and `findInvalidAncestorForTag` are [exposed in `react-dom@18.2.0/cjs/react-dom.development.js`](https://unpkg.com/browse/react-dom@18.2.0/cjs/react-dom.development.js)

**Comment by ljharb** [comment]
If we can access them from react, that’d be sufficient, but we’d have to be able to do so across many react versions :-/

**Comment by karlhorky** [comment]
Ah looks like there is an ESLint plugin from @MananTank that does this over here 👀

- https://github.com/MananTank/eslint-plugin-validate-jsx-nesting

**Comment by ljharb** [comment]
That's a great POC for a new rule! I'd love to see a PR that adds it, with lots of test cases :-D

**Comment by adeleke5140** [comment]
@ljharb Hello, I notice this PR is still open. what is required to get it to completion and merged into the package?

I ran into some `invalid dom nesting` errors recently and if `eslint` can catch that before runtime checks, it would be great.

**Comment by ljharb** [comment]
@adeleke5140 this isn’t a PR, so what’s required is “a PR” :-)

**Comment by karlhorky** [comment]
@adeleke5140 if you'd like to create a PR for this to get it into `eslint-plugin-react`, that would be amazing! 

Here are some potentially useful links:

1. Our simple version using `no-restricted-syntax` rules here: https://github.com/upleveled/eslint-config-upleveled/blob/8042b1ff73c1f7d322f80caf796257a4df86cb69/index.cjs#L66
2. Much more comprehensive [`eslint-plugin-validate-jsx-nesting`](https://github.com/MananTank/eslint-plugin-validate-jsx-nesting) by @MananTank
